### PR TITLE
[ENG-3186][ENG-3187][ENG-3188][ENG-3189]  Add ROR data to crossref and datacite, fix bugs

### PR DESCRIPTION
**Pull request by Johnetordoff**
## Purpose

We have got a bunch of tweaks and tune-ups for our metadata system, some of which effect our Institutions model.

### Institutions Model
- add management command to populate ROR data Institutions dict
- adds `ror` field to the institution model

### Preprints Serializer
- prevents user from adding their own preprint DOI as an `article_doi`

### Datacite
- adds to affiliation json building logic, to include names and domains
- add OSF as Hosting institution contributor
-  fixes issue where guid link was wrong
- add "Social" ORCID to metadata payload

### Crossref

- adds contributor-based affiliation and ROR data
- add "Social" ORCID to metadata payload as unverified

## Documentation

🐛 fix, no docs.

## Side Effects

None that I know of.

## Ticket

https://openscience.atlassian.net/browse/ENG-3186
https://openscience.atlassian.net/browse/ENG-3187
https://openscience.atlassian.net/browse/ENG-3188
https://openscience.atlassian.net/browse/ENG-3189

### Datacite
```json
{
  "id": "https://doi.org/10.70102/fk2osf.io/bs5ay",
  "doi": "10.70102/FK2OSF.IO/BS5AY",
  "url": "http://localhost:5000/bs5ay/",
  "types": {
    "ris": "RPRT",
    "bibtex": "article",
    "citeproc": "article-journal",
    "schemaOrg": "ScholarlyArticle",
    "resourceType": "Pre-registration",
    "resourceTypeGeneral": "Text"
  },
  "creators": [
    {
      "name": "Mercury1, Freddie",
      "nameType": "Personal",
      "givenName": "Freddie",
      "familyName": "Mercury1",
      "affiliation": [
        {
          "name": "California Policy Lab [Test]"
        },
        {
          "name": "Center for Astrophysics | Harvard & Smithsonian [Test]"
        },
        {
          "name": "Center for Astrophysics | Harvard & Smithsonian [Test]",
          "affiliationIdentifier": "https://ror.org/03c3r2d17",
          "affiliationIdentifierScheme": "ROR"
        },
        {
          "name": "Character Lab Research Network [Test]"
        }
      ],
      "nameIdentifiers": [
        {
          "nameIdentifier": "http://localhost:5000/2tgcj/",
          "nameIdentifierScheme": "URL"
        }
      ]
    }
  ],
  "titles": [
    {
      "title": "Networked bandwidth-monitored model"
    }
  ],
  "publisher": "Open Science Framework",
  "container": {},
  "subjects": [],
  "contributors": [
    {
      "name": "Mercury1, Freddie",
      "nameType": "Personal",
      "givenName": "Freddie",
      "familyName": "Mercury1",
      "affiliation": [
        {
          "name": "California Policy Lab [Test]"
        },
        {
          "name": "Center for Astrophysics | Harvard & Smithsonian [Test]"
        },
        {
          "name": "Center for Astrophysics | Harvard & Smithsonian [Test]",
          "affiliationIdentifier": "https://ror.org/03c3r2d17",
          "affiliationIdentifierScheme": "ROR"
        },
        {
          "name": "Character Lab Research Network [Test]"
        }
      ],
      "contributorType": "ProjectMember",
      "nameIdentifiers": [
        {
          "nameIdentifier": "http://localhost:5000/2tgcj/",
          "nameIdentifierScheme": "URL"
        }
      ]
    },
    {
      "name": "Open Science Framework",
      "nameType": "Organizational",
      "affiliation": [],
      "contributorType": "HostingInstitution",
      "nameIdentifiers": [
        {
          "nameIdentifier": "https://ror.org/05d5mza29/",
          "nameIdentifierScheme": "ROR"
        },
        {
          "nameIdentifier": "https://grid.ac/institutes/grid.466501.0/",
          "nameIdentifierScheme": "GRID"
        }
      ]
    }
  ],
  "dates": [
    {
      "date": "2021-10-11",
      "dateType": "Created"
    },
    {
      "date": "2021-10-11",
      "dateType": "Updated"
    },
    {
      "date": "2021-10-19",
      "dateType": "Issued"
    }
  ],
  "publicationYear": 2021,
  "identifiers": [],
  "sizes": [],
  "formats": [],
  "rightsList": [],
  "descriptions": [
    {
      "description": "Authority possible safe.",
      "descriptionType": "Abstract"
    }
  ],
  "geoLocations": [],
  "fundingReferences": [],
  "relatedIdentifiers": [],
  "relatedItems": [],
  "schemaVersion": "http://datacite.org/schema/kernel-4",
  "providerId": "demo",
  "clientId": "demo.cos",
  "agency": "datacite",
  "state": "findable"
}
```

### Crossref

```xml
<doi_batch xmlns="http://www.crossref.org/schema/5.3.1" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" version="5.3.1" xsi:schemaLocation="http://www.crossref.org/schema/5.3.1 http://www.crossref.org/schemas/crossref5.3.1.xsd">
   <head>
      <doi_batch_id>qyvkh</doi_batch_id>
      <timestamp>1634827941</timestamp>
      <depositor>
         <depositor_name>Open Science Framework</depositor_name>
         <email_address>john@cos.io</email_address>
      </depositor>
      <registrant>Center for Open Science</registrant>
   </head>
   <body>
      <posted_content type="preprint">
         <group_title>Open Science Framework</group_title>
         <contributors>
            <person_name sequence="first" contributor_role="author">
               <given_name>Freddie</given_name>
               <surname>Mercury0</surname>
               <affiliations>
                  <institution>
                     <institution_name>Temple University [Test]</institution_name>
                     <institution_id type="ror">https://ror.org/05d5mza29</institution_id>
                  </institution>
               </affiliations>
               <ORCID authenticated="false">https://orcid.org/0100-0101-5100-0107</ORCID>
            </person_name>
         </contributors>
         <titles>
            <title>test</title>
         </titles>
         <posted_date>
            <month>08</month>
            <day>03</day>
            <year>2021</year>
         </posted_date>
         <item_number>osf.io/qyvkh</item_number>
         <abstract xmlns="http://www.ncbi.nlm.nih.gov/JATS1">
            <p>bstract:bstract:bstract:bstract:bstract:</p>
         </abstract>
         <program xmlns="http://www.crossref.org/AccessIndicators.xsd">
            <license_ref start_date="2021-08-03">https://creativecommons.org/licenses/by/4.0/legalcode</license_ref>
         </program>
         <program xmlns="http://www.crossref.org/relations.xsd">
            <related_item>
               <intra_work_relation relationship-type="isPreprintOf" identifier-type="doi">10.70102/FK2osf.io/qyvkh</intra_work_relation>
            </related_item>
         </program>
         <doi_data>
            <doi>10.70102/FK2osf.io/qyvkh</doi>
            <resource>http://localhost:5000/qyvkh</resource>
         </doi_data>
      </posted_content>
   </body>
</doi_batch>


```

### CODE REVIEW

**Comment by jwalz** [comment]
The migration might need to be updated again, since there was a hotfix with a migration the other day

**Review by jwalz — COMMENTED** [review]
(most of these comments came up in 1:1)

File: osf/metadata/utils.py [hunk 1]
```diff
@@ -23,44 +23,81 @@
...
+    if user.external_identity.get('ORCID'):
+        verified = list(user.external_identity['ORCID'].values())[0] == 'VERIFIED'
+        if verified:
+            data['nameIdentifiers'].append({
+                'nameIdentifier': list(user.external_identity['ORCID'].keys())[0],
+                'nameIdentifierScheme': 'ORCID',
+                'schemeURI': 'http://orcid.org/'
+            })
+
+    if user.social_links.get('orcid'):
```
Comment (jwalz) [inline osf/metadata/utils.py:45]: only update if verified

File: osf/metadata/utils.py [hunk 2]
```diff
@@ -23,44 +23,81 @@
...
+    return data
+
+
+def datacite_format_affilations(user):
+    data = {'affiliation': []}
+    for affiliated_institution in user.affiliated_institutions.all():
+        data['affiliation'].append({
+            'name': affiliated_institution.name,
+        })
+        for domain in affiliated_institution.domains:
```
Comment (jwalz) [inline osf/metadata/utils.py:61]: short term: remove this, as these domains seem OSF-specific. Ticket to be made for resolving how to store actual institutional domains for metadata

File: website/identifiers/clients/datacite.py [hunk 3]
```diff
@@ -33,31 +33,58 @@ def build_metadata(self, node):
...
-            },
+            'identifiers': [
+                {
+                    'identifier': self.build_doi(node),
+                    'identifierType': 'DOI',
+                }
+            ],
             'creators': datacite_format_creators([node.creator]),
-            'contributors': datacite_format_contributors(node.visible_contributors),
+            'contributors': contributors,
```
Comment (jwalz) [inline website/identifiers/clients/datacite.py:64]: This should be non-bibliographic contribs with bibliographic contribs going in `creators`: https://blog.datacite.org/contributor-information-in-datacite-metadata/

**Review by jwalz — CHANGES_REQUESTED** [review]
A few changes

File: osf/metadata/utils.py (hunk 1 above)
Comment (jwalz) [reply osf/metadata/utils.py:45]: (and don't send unverified to Crossref)

File: website/identifiers/clients/datacite.py [hunk 4]
```diff
@@ -33,31 +33,58 @@ def build_metadata(self, node):
         """Return the formatted datacite metadata XML as a string.
          """
 
+        contributors = datacite_format_contributors(node.contributors)
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply website/identifiers/clients/datacite.py:36]: I'm not sure, looking at [this example](https://schema.datacite.org/meta/kernel-4.3/example/datacite-example-full-v4.xml) the creator is a Datacite employee (the creator of the metadata record), not the the actual contributors, this seems to clash with what Eric said. I assumed he meant anyone listed is a contributor. This example is more aligned my original behavior, but lets confirm with Eric.

[the remark this answers is not in the document]
Comment (jwalz) [reply website/identifiers/clients/datacite.py:36]: poke: still need to exclude bibliographic contribs here, right?

**Review by jwalz — APPROVED** [review]

**Pull request merged**